torch/csrc/autograd/input_buffer.cpp

```diff
@@ -257,6 +272,11 @@ void InputBuffer::add(
     event.record(*opt_producer_stream);
     ready_events[pos] = std::move(event);
     ready_streams[pos] = opt_producer_stream;
+    if (autograd_avoid_record_streams() && record_stream_used) {
+      auto event_for_record_stream = c10::Event{device_type};
+      event.record(*opt_consumer_stream);
```

Comment:
I think you actually want to have the producer stream wait for consumer stream AFTER the consumer is done using the gradient

e.g. record on the InputBuffer the stream on which the gradient Tensor was allocated, and have that stream wait for the consumer stream after the node finishes computing but before the gradient it is freed.